native/src/main/java/io/ballerina/stdlib/serdes/SchemaGenerator.java

```diff
@@ -43,201 +75,538 @@
+            String fieldName,
+            int fieldNumber) {
+
+        String protoType = DataTypeMapper.mapBallerinaTypeToProtoType(ballerinaType.getTag());
+        ProtobufMessageFieldBuilder messageField = new ProtobufMessageFieldBuilder(
+                OPTIONAL_LABEL,
+                protoType,
+                fieldName,
+                fieldNumber);
+        messageBuilder.addField(messageField);
     }
 
-    private static void buildProtobufMessageForBallerinaArrayType(ProtobufMessageBuilder messageBuilder,
-                                                                  ArrayType arrayType, String name, int number,
-                                                                  int unionFieldIdentifier) {
-        Type type = arrayType.getElementType();
-
-        if (type.getTag() == TypeTags.UNION_TAG) {
-            String fieldName = name + "_" + UNION_TYPE_IDENTIFIER;
-            messageBuilder.addNestedMessage(buildProtobufMessageForBallerinaUnionType(type, fieldName));
-            messageBuilder.addField(REPEATED_LABEL, fieldName, name, number);
-        } else if (type.getTag() <= TypeTags.BOOLEAN_TAG) {
-            String protoElementType = DataTypeMapper.getProtoTypeFromTag(type.getTag());
-            if (protoElementType.equals(BYTES)) {
-                messageBuilder.addField(OPTIONAL_LABEL, protoElementType, name, number);
-            } else {
-                messageBuilder.addField(REPEATED_LABEL, protoElementType, name, number);
-            }
-        } else if (type.getTag() == TypeTags.ARRAY_TAG) {
-            ArrayType nestedArrayType = (ArrayType) type;
-            String nestedMessageName;
-            int elementTag = nestedArrayType.getElementType().getTag();
-            if (elementTag == TypeTags.UNION_TAG || elementTag == TypeTags.ARRAY_TAG) {
-                nestedMessageName = NESTED_UNION_FIELD_NAME + unionFieldIdentifier;
-                unionFieldIdentifier++;
-            } else {
-                nestedMessageName = nestedArrayType.getElementType().getName();
+    // Generates schema for ballerina decimal type
+    private static void generateMessageDefinitionForPrimitiveDecimal(ProtobufMessageBuilder messageBuilder) {
+
+        int fieldNumber = 1;
+
+        // Java BigDecimal representation used for serializing ballerina decimal value
+        var scaleField = new ProtobufMessageFieldBuilder(OPTIONAL_LABEL, UINT32, SCALE, fieldNumber++);
+        var precisionField = new ProtobufMessageFieldBuilder(OPTIONAL_LABEL, UINT32, PRECISION, fieldNumber++);
+        var valueField = new ProtobufMessageFieldBuilder(OPTIONAL_LABEL, BYTES, VALUE, fieldNumber);
+
+        messageBuilder.addField(scaleField);
+        messageBuilder.addField(precisionField);
+        messageBuilder.addField(valueField);
+    }
+
+    private static Map.Entry<String, Type> mapUnionMemberToMapEntry(Type type) {
+
+        String typeName = type.getName();
+        if (type instanceof ArrayType) {
```

Comment:
When checking Ballerina Type objects, instead of `instanceof` check, use the type tag. Since it is just int comparison, this is considered faster than compared to the `instanceof` check. Check other places as well.

```suggestion
        if (type.getTag() == TypeTags.ARRAY_TAG) {
```

Reply:
Noted

Reply:
Updated all